### PR TITLE
Single pixel buffer

### PR DESCRIPTION
**Pull request by cmeissl**

### CODE REVIEW

File: src/backend/renderer/element/surface.rs [hunk 1]
```diff
@@ -266,33 +270,51 @@ where
...
 
         let view = state.view()?;
         let buffer = state.buffer()?;
+        let buffer_type = crate::backend::renderer::buffer_type(buffer)?;
         let id = Id::from_wayland_resource(surface);
 
-        let texture = state.texture::<R>(renderer.id())?;
+        if matches!(buffer_type, crate::backend::renderer::BufferType::SinglePixel) {
+            let spb = crate::wayland::single_pixel_buffer::get_single_pixel_buffer(buffer).ok()?;
+            let rgba = spb.rgba8888();
```
Comment (Drakulix) [inline src/backend/renderer/element/surface.rs:306]: So we truncate it to rgba8888, just to convert it to floats afterwards..

Can we add a `rgba16f` variant please?

**Review by Drakulix — APPROVED** [review]
Good start. I really like the re-use of the TextureRenderElement.

File: src/backend/renderer/element/surface.rs (hunk 1 above)
Comment (cmeissl) [reply src/backend/renderer/element/surface.rs:306]: yeah, that is kinda pointless. any reason why to use half float? After reading the spec (and looking at what swaybg sends) the values range from 0 - u32::MAX. So dividing by `u32::MAX as f32` (or just `0xffffffff`) should give us `f32` per component?

**Review by bwidawsk — APPROVED** [review]
LGTM, and I'm currently using this in my compositor.

**Comment by bwidawsk** [comment]
Is there anything needed to get this merged?

File: src/wayland/single_pixel_buffer/handlers.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                id: buffer,
+                r,
+                g,
+                b,
+                a,
+            } => {
+                data_init.init(buffer, SinglePixelBufferUserData { r, g, b, a });
+            }
+            wp_single_pixel_buffer_manager_v1::Request::Destroy => {}
+            _ => todo!(),
```
Comment (chrisduerr) [inline src/wayland/single_pixel_buffer/handlers.rs:56]: `todo!()` shouldn't be used in this fashion, you should probably use `unimplemented!()` instead.

**Review by chrisduerr — CHANGES_REQUESTED** [review]
+1 to compositors implementing this protocol using this PR:
https://github.com/catacombing/catacomb/commit/595eb4d9f5f4636e5d8783463b76bfb44fb84d5c

I think it would be nice to have this merged upstream soon since it requires some changes to Smithay methods (buffer_dimensions/buffer_type) and without these wrappers need to be used downstream.

**Pull request closed without merging**